src/neighbors.cc

```diff
@@ -1275,32 +1275,35 @@ peerRefreshDNS(void *data)
     eventAddIsh("peerRefreshDNS", peerRefreshDNS, nullptr, 3600.0, 1);
 }
 
-static void
-peerConnectFailedSilent(CachePeer * p)
+// TODO: Move to CachePeer::noteFailure() or similar.
+// TODO: Require callers to detail failures instead of using one (and often
+// misleading!) "TCP" label for all of them.
+void
+peerConnectFailed(CachePeer * const p)
 {
     p->stats.last_connect_failure = squid_curtime;
+    if (p->tcp_up > 0)
+        --p->tcp_up;
 
-    if (!p->tcp_up) {
-        debugs(15, 2, "TCP connection to " << p->host << "/" << p->http_port <<
-               " dead");
-        return;
-    }
+    // TODO: Report peer name. Same-addresses peers often have different names.
```

Comment:
We actually wasted some time during triage because the admin thought that Squid reports unexpected multiple failures for the same cache_peer. Turns out they were expected single/isolated failures associated with different same-address cache_peers (which is a perfectly valid configuration).

I can address this TODO in this PR if anybody insists on that, but doing so will change level-1 messages that will be changed again when we start detailing these failures (see the other TODOs above) and will probably trigger changes beyond this function. I think it is best not to address this problem in this PR.

Reply:
What do you mean by this exactly? the p->name is already reported on L#1296:

>            debugs(15, DBG_IMPORTANT, "Detected DEAD " << neighborTypeStr(p) << ": " << p->name);


AFAICS, altering the "is still DEAD" debugs output to include the p->name is all that is necessary to meet this TODO.

Reply:
> What do you mean by this exactly?

Sorry, I do not know which "this" you are asking about.

> the p->name is already reported on L#1296:

... but not in some other contexts. Here are all the ones I could find quickly (citing here their PR variants, but these or equivalent debugs() are present in the official code as well):

```C++
debugs(38, 3, "netdbExchangeHandleReply: for '" << ex->p->host << ":" << ex->p->http_port << "'");
debugs(15, Important(29), "Configuring " << neighborTypeStr(p) << " " << p->host << "/" << p->http_port << "/" << p->icp.port);
debugs(15, level, "ERROR: TCP connection to " << p->host << "/" << p->http_port << " failed");
debugs(15, 2, "cache_peer " << p->host << "/" << p->http_port << " is still DEAD");
debugs(15, 2, "TCP connection to " << p->host << "/" << p->http_port << " succeeded");
debugs(72, 3, "peerDigestCheck: peer " <<  pd->peer->host << ":" << pd->peer->http_port);
debugs(15, DBG_IMPORTANT, "WARNING: Peer looks like this host." <<
       Debug::Extra << "Ignoring " <<
       neighborTypeStr(thisPeer) << " " << thisPeer->host <<
       "/" << thisPeer->http_port << "/" <<
       thisPeer->icp.port);
```

Ideally, a dedicated PR should make CachePeer reporting consistent and convenient while avoid code duplication and other problems. I volunteer to post such a PR. However, again, if you insist that we do this out-of-scope polishing here, I will do it here.